blockchain/process.go

```diff
@@ -129,6 +131,137 @@ func (b *BlockChain) processOrphans(hash *chainhash.Hash, flags BehaviorFlags) e
 	return nil
 }
 
+// checkKnownInvalidBlock returns an appropriate error when the provided block
+// is known to be invalid either due to failing validation itself or due to
+// having a known invalid ancestor (aka being part of an invalid branch).
+//
+// This function is safe for concurrent access.
+func (b *BlockChain) checkKnownInvalidBlock(node *blockNode) error {
+	status := b.index.NodeStatus(node)
+	if status.KnownValidateFailed() {
+		str := fmt.Sprintf("block %s is known to be invalid", node.hash)
+		return ruleError(ErrKnownInvalidBlock, str)
+	}
+	if status.KnownInvalidAncestor() {
+		str := fmt.Sprintf("block %s is known to be part of an invalid branch",
+			node.hash)
+		return ruleError(ErrInvalidAncestorBlock, str)
+	}
+
+	return nil
+}
+
+// maybeAcceptBlockHeader potentially accepts the header to the block index and,
+// if accepted, returns the block node associated with the header.  It performs
+// several context independent checks as well as those which depend on its
+// position within the chain.  It should be noted that some of the header fields
+// require the full block data to be available in order to be able to validate
+// them, so those fields are not included here.  This provides support for full
+// headers-first semantics.
+//
+// The flag for check header sanity allows the additional header sanity checks
+// to be skipped which is useful for the full block processing path which checks
+// the sanity of the entire block, including the header, before attempting to
+// accept its header in order to quickly eliminate blocks that are obviously
+// incorrect.
+//
+// In the case the block header is already known, the associated block node is
+// examined to determine if the block is already known to be invalid, in which
+// case an appropriate error will be returned.  Otherwise, the block node is
+// returned.
+//
+// This function MUST be called with the chain lock held (for writes).
+func (b *BlockChain) maybeAcceptBlockHeader(header *wire.BlockHeader, checkHeaderSanity bool) (*blockNode, error) {
```

Comment:
I'd prefer putting this function in `accept.go`

Reply:
changed the location of the functions